File: games/lynrummy/python/move.py
"""
move.py — BFS desc rendering: describe_move + narrate + hint.

The Python equivalent of `Game.Agent.Move.elm`'s rendering
half. The desc dicts themselves are still inline literals
built by the enumerator (Elm has typed records); see
ALIGNMENT_REPORT.md § 3 for the dataclass opportunity that
hasn't been taken yet.

Lifted from `bfs_solver.py` 2026-04-26 as the module split
landed.
"""

from cards import classify, label_d


def stack_label(stack):
    return " ".join(label_d(c) for c in stack)


def narrate(desc):
    """Evocative one-liner for a move, communicating INTENT
    rather than mechanics. Steve-facing: this is how Claude
    narrates what the agent is doing in the verbose-mode
    log. Each move type gets a verb-forward phrasing at the
    human chunk level (engulf, splice, pop, tuck, ...).

    For exact structural matching, use `describe_move`. For
    the vague hint a human PLAYER would see in the UI, use
    `hint`.
    """
    if desc["type"] == "free_pull":
        loose = label_d(desc["loose"])
        result = stack_label(desc["result"])
        check = " ✓" if desc["graduated"] else ""
        return f"pull {loose} into [{result}]{check}"

    if desc["type"] == "extract_absorb":
        verb = desc["verb"]
        ec = label_d(desc["ext_card"])
        result = stack_label(desc["result"])
        check = " ✓" if desc["graduated"] else ""
        spawned = ""
        if desc["spawned"]:
            spawned = (" (leaves "
                       + ", ".join("[" + stack_label(s) + "]"
                                   for s in desc["spawned"])
                       + " homeless)")
        return f"{verb} {ec} → [{result}]{check}{spawned}"

    if desc["type"] == "shift":
        p = label_d(desc["p_card"])
        stolen = label_d(desc["stolen"])
        merged = stack_label(desc["merged"])
        check = " ✓" if desc["graduated"] else ""
        return f"{p} pops {stolen} → [{merged}]{check}"

    if desc["type"] == "splice":
        loose = label_d(desc["loose"])
        left = stack_label(desc["left_result"])
        right = stack_label(desc["right_result"])
        return f"splice {loose} → [{left}] + [{right}]"

    if desc["type"] == "push":
        tb = stack_label(desc["trouble_before"])
        target = stack_label(desc["target_before"])
        result = stack_label(desc["result"])
        # Engulf-shape vs plain push: plain push extends a
        # helper by 1-2 cards; engulf swallows a helper into a
        # complete stack (graduated from GROWING).
        if classify(desc["result"]) != "other":
            return f"engulf [{target}] into [{tb}] → [{result}] ✓"
        return f"tuck [{tb}] into [{target}] → [{result}]"

    return str(desc)


def hint(desc):
    """Vague-but-useful one-liner for a HUMAN PLAYER. Names
    the verb + a key card/partial + (sometimes) the GROUP
    KIND of the destination, but does NOT spell out the
    specific source/target stacks. The intent: nudge without
    solving.

    Steve's reference phrasing: "You can splice the 7H into a
    red-black run." (Names card, verb, group kind. Doesn't
    name the run.)
    """
    if desc["type"] == "free_pull":
        loose = label_d(desc["loose"])
        kind = group_kind_phrase(desc["result"])
        return f"You can pull the {loose} onto {kind}."

    if desc["type"] == "extract_absorb":
        verb = desc["verb"]
        ec = label_d(desc["ext_card"])
        target_kind = partial_kind_phrase(desc["target_before"])
        return f"You can {verb} the {ec} to extend {target_kind}."

    if desc["type"] == "shift":
        p = label_d(desc["p_card"])
        stolen = label_d(desc["stolen"])
        return (f"You can pop the {stolen} by shifting "
                f"the {p} into the run.")

    if desc["type"] == "splice":
        loose = label_d(desc["loose"])
        # Source is always a length-4+ pure or rb run.
        run_kind = run_kind_phrase(desc["source"])
        return f"You can splice the {loose} into a {run_kind}."

    if desc["type"] == "push":
        tb = stack_label(desc["trouble_before"])
        if classify(desc["result"]) != "other":
            return f"You can complete a run by absorbing [{tb}]."
        return f"You can tuck [{tb}] back into a run."

    return None


def group_kind_phrase(stack):
    """Render the GROUP KIND of a (presumably legal) stack as
    natural language: 'a red-black run', 'a set', etc."""
    kind = classify(stack)
    if kind == "set":
        return "a set"
    if kind == "pure_run":
        return "a pure run"
    if kind == "rb_run":
        return "a red-black run"
    return "a partial"


def partial_kind_phrase(stack):
    """For a 1- or 2-card target (trouble singleton or
    growing 2-partial). Calls out the headline card."""
    n = len(stack)
    if n == 0:
        return "an empty target"
    if n == 1:
        return f"the {label_d(stack[0])}"
    return ("the partial ["
            + " ".join(label_d(c) for c in stack)
            + "]")


def run_kind_phrase(stack):
    """For a length-4+ run (used by splice). Says 'pure run'
    or 'red-black run'."""
    kind = classify(stack)
    if kind == "pure_run":
        return "pure run"
    if kind == "rb_run":
        return "red-black run"
    return "run"


def describe_move(desc):
    """Render a one-line DSL string for a move."""
    if desc["type"] == "free_pull":
        loose = label_d(desc["loose"])
        bucket = desc["target_bucket_before"]
        tb = stack_label(desc["target_before"])
        result = stack_label(desc["result"])
        graduated = " [→COMPLETE]" if desc["graduated"] else ""
        return (f"pull {loose} onto {bucket} [{tb}] → "
                f"[{result}]{graduated}")
    if desc["type"] == "extract_absorb":
        verb = desc["verb"]
        ec = label_d(desc["ext_card"])
        src = stack_label(desc["source"])
        bucket = desc["target_bucket_before"]
        tb = stack_label(desc["target_before"])
        result = stack_label(desc["result"])
        spawned = ""
        if desc["spawned"]:
            spawned = (" ; spawn TROUBLE: "
                       + ", ".join("[" + stack_label(s) + "]"
                                   for s in desc["spawned"]))
        graduated = " [→COMPLETE]" if desc["graduated"] else ""
        return (f"{verb} {ec} from HELPER [{src}], "
                f"absorb onto {bucket} [{tb}] → "
                f"[{result}]{graduated}{spawned}")
    if desc["type"] == "shift":
        p = label_d(desc["p_card"])
        stolen = label_d(desc["stolen"])
        new_donor = stack_label(desc["new_donor"])
        new_source = desc["new_source"]
        p_idx = new_source.index(desc["p_card"])
        rest = [c for c in new_source if c != desc["p_card"]]
        rest_label = " ".join(label_d(c) for c in rest)
        if p_idx == 0:
            shifted = f"{p} + {rest_label}"
        else:
            shifted = f"{rest_label} + {p}"
        bucket = desc["target_bucket_before"]
        tb = stack_label(desc["target_before"])
        merged = stack_label(desc["merged"])
        graduated = " [→COMPLETE]" if desc["graduated"] else ""
        return (f"shift {p} to pop {stolen} "
                f"[{new_donor} -> {shifted}]; "
                f"absorb onto {bucket} [{tb}] → "
                f"[{merged}]{graduated}")
    if desc["type"] == "splice":
        loose = label_d(desc["loose"])
        src = stack_label(desc["source"])
        left = stack_label(desc["left_result"])
        right = stack_label(desc["right_result"])
        return (f"splice [{loose}] into HELPER [{src}] → "
                f"[{left}] + [{right}]")
    if desc["type"] == "push":
        tb = stack_label(desc["trouble_before"])
        target = stack_label(desc["target_before"])
        result = stack_label(desc["result"])
        return (f"push TROUBLE [{tb}] onto HELPER [{target}] → "
                f"[{result}]")
    return str(desc)

File: games/lynrummy/ts/src/move.ts
// move.ts — BFS descriptor types + plan-line rendering.
//
// TS port of python/move.py. Each move type has a dedicated descriptor
// shape mirroring the python dataclass field set. The enumerator emits
// these; readers dispatch on the `type` discriminator.
//
// Descriptors carry RAW card tuples / card lists, NOT CCS objects, for
// plan-line stability and downstream serialization. Mirroring the
// python field layout exactly keeps the cross-language plan-line check
// trivial (string equality, no normalization).

import type { Card } from "./rules/card.ts";
import { RANKS, SUITS, cardLabel } from "./rules/card.ts";

/** Side discriminator for absorb/splice variants. Mirrors python's
 *  `Side`. Only data-layout-level — no function takes `side` as a
 *  parameter (per SOLVER.md's discipline). */
export type Side = "left" | "right";

/** Source verb that produced an extract. Mirrors python's `Verb`. */
export type Verb = "peel" | "pluck" | "yank" | "steal" | "split_out";

/** Bucket where the absorb target came from. Same alphabet as
 *  `BucketName` in buckets.ts. */
export type AbsorberBucket = "trouble" | "growing";

// --- Descriptors -----------------------------------------------------------
//
// Each interface mirrors the python dataclass field-for-field. The
// `type` literal field discriminates them.

export interface ExtractAbsorbDesc {
  readonly type: "extract_absorb";
  readonly verb: Verb;
  readonly source: readonly Card[];
  readonly extCard: Card;
  readonly targetBefore: readonly Card[];
  readonly targetBucketBefore: AbsorberBucket;
  readonly result: readonly Card[];
  readonly side: Side;
  readonly graduated: boolean;
  readonly spawned: readonly (readonly Card[])[];
}

export interface FreePullDesc {
  readonly type: "free_pull";
  readonly loose: Card;
  readonly targetBefore: readonly Card[];
  readonly targetBucketBefore: AbsorberBucket;
  readonly result: readonly Card[];
  readonly side: Side;
  readonly graduated: boolean;
}

export interface PushDesc {
  readonly type: "push";
  readonly troubleBefore: readonly Card[];
  readonly targetBefore: readonly Card[];
  readonly result: readonly Card[];
  readonly side: Side;
}

export interface SpliceDesc {
  readonly type: "splice";
  readonly loose: Card;
  readonly source: readonly Card[];
  readonly k: number;
  readonly side: Side;
  readonly leftResult: readonly Card[];
  readonly rightResult: readonly Card[];
}

export interface ShiftDesc {
  readonly type: "shift";
  readonly source: readonly Card[];
  readonly donor: readonly Card[];
  readonly stolen: Card;
  readonly pCard: Card;
  readonly whichEnd: number;
  readonly newSource: readonly Card[];
  readonly newDonor: readonly Card[];
  readonly targetBefore: readonly Card[];
  readonly targetBucketBefore: AbsorberBucket;
  readonly merged: readonly Card[];
  readonly side: Side;
  readonly graduated: boolean;
}

/** Decompose: split a TROUBLE pair (pair_run / pair_rb / pair_set)
 *  back into two singletons. The bundling that pair-spawning moves
 *  produce isn't a real game commitment — sometimes the right play
 *  separates the cards. See `random233.md`. */
export interface DecomposeDesc {
  readonly type: "decompose";
  readonly pairBefore: readonly Card[];   // 2-card pair being split
  readonly leftCard: Card;
  readonly rightCard: Card;
}

/** Discriminated union over all move descriptors. */
export type Desc =
  | ExtractAbsorbDesc
  | FreePullDesc
  | PushDesc
  | SpliceDesc
  | ShiftDesc
  | DecomposeDesc;

// --- Plan-line rendering ---------------------------------------------------
//
// `describe(desc)` produces the canonical one-line DSL string. The
// strings here are the cross-language contract — they must exactly
// match python's `move.describe(desc)` output.

function stackLabel(stack: readonly Card[]): string {
  return stack.map(cardLabel).join(" ");
}

/**
 * Render a one-line DSL string for a move. Mirrors python's
 * `move.describe`. Must match python output exactly — this is the
 * cross-language contract.
 */
export function describe(desc: Desc): string {
  switch (desc.type) {
    case "free_pull": {
      const graduated = desc.graduated ? " [→COMPLETE]" : "";
      return `pull ${cardLabel(desc.loose)} onto ${desc.targetBucketBefore} `
        + `[${stackLabel(desc.targetBefore)}] → `
        + `[${stackLabel(desc.result)}]${graduated}`;
    }
    case "extract_absorb": {
      let spawnedStr = "";
      if (desc.spawned.length > 0) {
        spawnedStr = " ; spawn TROUBLE: "
          + desc.spawned.map(s => "[" + stackLabel(s) + "]").join(", ");
      }
      const graduated = desc.graduated ? " [→COMPLETE]" : "";
      return `${desc.verb} ${cardLabel(desc.extCard)} from HELPER `
        + `[${stackLabel(desc.source)}], `
        + `absorb onto ${desc.targetBucketBefore} `
        + `[${stackLabel(desc.targetBefore)}] → `
        + `[${stackLabel(desc.result)}]`
        + `${graduated}${spawnedStr}`;
    }
    case "shift": {
      const p = cardLabel(desc.pCard);
      // Determine whether p ended up at index 0 or last (Python uses
      // .index(p_card) on new_source then enumerates rest in order).
      let pIdx = -1;
      for (let i = 0; i < desc.newSource.length; i++) {
        const c = desc.newSource[i]!;
        if (c[0] === desc.pCard[0] && c[1] === desc.pCard[1] && c[2] === desc.pCard[2]) {
          pIdx = i;
          break;
        }
      }
      const rest: Card[] = [];
      for (const c of desc.newSource) {
        if (!(c[0] === desc.pCard[0] && c[1] === desc.pCard[1] && c[2] === desc.pCard[2])) {
          rest.push(c);
        }
      }
      const restLabel = rest.map(cardLabel).join(" ");
      const shifted = pIdx === 0
        ? `${p} + ${restLabel}`
        : `${restLabel} + ${p}`;
      const graduated = desc.graduated ? " [→COMPLETE]" : "";
      return `shift ${p} to pop ${cardLabel(desc.stolen)} `
        + `[${stackLabel(desc.newDonor)} -> ${shifted}]; `
        + `absorb onto ${desc.targetBucketBefore} `
        + `[${stackLabel(desc.targetBefore)}] → `
        + `[${stackLabel(desc.merged)}]${graduated}`;
    }
    case "splice": {
      return `splice [${cardLabel(desc.loose)}] into HELPER `
        + `[${stackLabel(desc.source)}] → `
        + `[${stackLabel(desc.leftResult)}] + `
        + `[${stackLabel(desc.rightResult)}]`;
    }
    case "push": {
      return `push TROUBLE [${stackLabel(desc.troubleBefore)}] onto HELPER `
        + `[${stackLabel(desc.targetBefore)}] → `
        + `[${stackLabel(desc.result)}]`;
    }
    case "decompose": {
      return `decompose TROUBLE [${stackLabel(desc.pairBefore)}] → `
        + `[${cardLabel(desc.leftCard)}] + [${cardLabel(desc.rightCard)}]`;
    }
  }
}

// --- Narrate / hint --------------------------------------------------------
//
// These are convenience renderers used by enumerate_moves conformance
// scenarios that assert `narrate_contains` or `hint_contains`. The
// outputs only need to MATCH on the substring — full output equality
// against python is NOT required (different code paths consume them).

/** Family-kind classifier used for narrate/hint phrasing. Returns
 *  "set" / "pure_run" / "rb_run" / "other". A simplified version of
 *  python's rules.classify; only correct on length-3+ legal stacks. */
function classifyFamily(stack: readonly Card[]): string {
  const n = stack.length;
  if (n < 3) return "other";
  const c0 = stack[0]!;
  const c1 = stack[1]!;
  // Set: same value, distinct suits.
  if (c0[0] === c1[0]) {
    const seen = new Set<number>();
    for (const c of stack) {
      if (c[0] !== c0[0]) return "other";
      if (seen.has(c[1])) return "other";
      seen.add(c[1]);
    }
    return "set";
  }
  // Run: successive values, same-suit (pure) or alternating-color (rb).
  const succ = (v: number) => v === 13 ? 1 : v + 1;
  if (succ(c0[0]) !== c1[0]) return "other";
  const RED = new Set([1, 3]);
  if (c0[1] === c1[1]) {
    let prevV = c1[0];
    for (let i = 2; i < n; i++) {
      const c = stack[i]!;
      if (c[0] !== succ(prevV) || c[1] !== c0[1]) return "other";
      prevV = c[0];
    }
    return "pure_run";
  }
  if (RED.has(c0[1]) === RED.has(c1[1])) return "other";
  let prevV = c1[0];
  let prevRed = RED.has(c1[1]);
  for (let i = 2; i < n; i++) {
    const c = stack[i]!;
    if (c[0] !== succ(prevV)) return "other";
    const cRed = RED.has(c[1]);
    if (cRed === prevRed) return "other";
    prevV = c[0];
    prevRed = cRed;
  }
  return "rb_run";
}

function groupKindPhrase(stack: readonly Card[]): string {
  const k = classifyFamily(stack);
  if (k === "set") return "a set";
  if (k === "pure_run") return "a pure run";
  if (k === "rb_run") return "a red-black run";
  return "a partial";
}

function partialKindPhrase(stack: readonly Card[]): string {
  const n = stack.length;
  if (n === 0) return "an empty target";
  if (n === 1) return `the ${cardLabel(stack[0]!)}`;
  return "the partial [" + stack.map(cardLabel).join(" ") + "]";
}

function runKindPhrase(stack: readonly Card[]): string {
  const k = classifyFamily(stack);
  if (k === "pure_run") return "pure run";
  if (k === "rb_run") return "red-black run";
  return "run";
}

/** Evocative one-liner for a move (intent over mechanics). Mirrors
 *  python's `move.narrate` — used by conformance `narrate_contains`. */
export function narrate(desc: Desc): string {
  switch (desc.type) {
    case "free_pull": {
      const check = desc.graduated ? " ✓" : "";
      return `pull ${cardLabel(desc.loose)} into [${stackLabel(desc.result)}]${check}`;
    }
    case "extract_absorb": {
      const check = desc.graduated ? " ✓" : "";
      let spawnedStr = "";
      if (desc.spawned.length > 0) {
        spawnedStr = " (leaves "
          + desc.spawned.map(s => "[" + stackLabel(s) + "]").join(", ")
          + " homeless)";
      }
      return `${desc.verb} ${cardLabel(desc.extCard)} → `
        + `[${stackLabel(desc.result)}]${check}${spawnedStr}`;
    }
    case "shift": {
      const check = desc.graduated ? " ✓" : "";
      return `${cardLabel(desc.pCard)} pops ${cardLabel(desc.stolen)} → `
        + `[${stackLabel(desc.merged)}]${check}`;
    }
    case "splice":
      return `splice ${cardLabel(desc.loose)} → `
        + `[${stackLabel(desc.leftResult)}] + [${stackLabel(desc.rightResult)}]`;
    case "push": {
      // Engulf-shape vs plain push: engulf produces a length-3+ legal
      // group from a growing-style merge.
      if (classifyFamily(desc.result) !== "other") {
        return `engulf [${stackLabel(desc.targetBefore)}] into `
          + `[${stackLabel(desc.troubleBefore)}] → `
          + `[${stackLabel(desc.result)}] ✓`;
      }
      return `tuck [${stackLabel(desc.troubleBefore)}] into `
        + `[${stackLabel(desc.targetBefore)}] → `
        + `[${stackLabel(desc.result)}]`;
    }
    case "decompose":
      return `decompose [${stackLabel(desc.pairBefore)}] into singletons`;
  }
}

/** Vague hint for a HUMAN PLAYER. Mirrors python's `move.hint`. May
 *  return null (e.g., extract_absorb verbs that are too specific). */
export function hint(desc: Desc): string | null {
  switch (desc.type) {
    case "free_pull":
      return `You can pull the ${cardLabel(desc.loose)} onto `
        + `${groupKindPhrase(desc.result)}.`;
    case "extract_absorb":
      return `You can ${desc.verb} the ${cardLabel(desc.extCard)} to `
        + `extend ${partialKindPhrase(desc.targetBefore)}.`;
    case "shift":
      return `You can pop the ${cardLabel(desc.stolen)} by shifting `
        + `the ${cardLabel(desc.pCard)} into the run.`;
    case "splice":
      return `You can splice the ${cardLabel(desc.loose)} into a `
        + `${runKindPhrase(desc.source)}.`;
    case "push":
      if (classifyFamily(desc.result) !== "other") {
        return `You can complete a run by absorbing [${stackLabel(desc.troubleBefore)}].`;
      }
      return `You can tuck [${stackLabel(desc.troubleBefore)}] back into a run.`;
    case "decompose":
      return `You can split the [${stackLabel(desc.pairBefore)}] pair apart.`;
  }
}
